### ISSUE TITLE
[unbound-method] `this` annotation should fix error

### ISSUE THREAD

**Issue by ThomasdenH** [typescript-eslint/typescript-eslint#854]
**Repro**

```JSON
{
  "@typescript-eslint/unbound-method": [
    "error",
    {
      "ignoreStatic": true
    }
  ]
}
```

```TS
const obj = {
  func(this: void): void {
    console.log('a');
  }
};

function acceptsFunc(a: () => void) {
  a();
}

acceptsFunc(obj.func);
```

**Expected Result**
Since the method explicitely doesn't use `this` in the function, the rule should not warn. ([See also the relevant section in the Handbook](https://www.typescriptlang.org/docs/handbook/functions.html#this))

**Actual Result**

**Additional Info**
```
error    Avoid referencing unbound methods which may cause unintentional scoping of `this`  @typescript-eslint/unbound-method
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.0.0` |
| `@typescript-eslint/parser`        | `2.0.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.1.0` |
| `node`                             | `11.15.0` |
| `npm`                              | `6.7.0` |

**Comment by ryota-ka** [comment]
Hi, could I work on this? (In fact my implementation is almost ready.)

@bradzacher BTW is it preferred to add a new rule option?
It looks just like a false positive to me.

**Comment by bradzacher** [comment]
Feel free to raise a PR. No need for an option.
